### ISSUE TITLE
Node fails to gracefully terminate after decommissioning

### ISSUE THREAD

**Issue by carloruiz** [cockroachdb/cockroach#111620]
**Describe the problem**
Node shutdown hangs after node has been decommissioned.

Cluster configuration: 6 node cluster, not running a workload. No open connections to the CRDB node. `server.shutdown.connection_wait='25s'`, `server.shutdown.connection_wait='30m'`, `server.shutdown.query_wait=90s`. 

1. Run `cockroach node decommission <node-id>`
2. Decommission completes successfully.
3. Send SIGTERM to CockroachDB process. 
4. The node will not terminate until it is forcefully killed by Kubernetes 45m later once the grace period has transpired. Node shutdown seems to get stuck before it reaches the lease transfer phase. 

**To Reproduce**
See above.

**Expected behavior**
The CRDB process should almost immediately terminate on SIGTERM once the node has been decommissioned. 

**Additional data / screenshots**
Logs are attached

**Environment:**
 - CockroachDB version: 23.1.7
 - Server OS: Red Hat Enterprise Linux 8.8

**Additional context**
Important to note that node shutdown functions as expected during a node restart (a `drain` followed by SIGTERM). The problem only surfaces after node decommissioning. Impact includes: extremely slow operations, hypothetical SQL level impact for distributed statements.

Jira issue: CRDB-31990

**Comment by carloruiz** [comment]
[termination_logs.zip](https://github.com/cockroachdb/cockroach/files/12786768/termination_logs.zip)

**Comment by rafiss** [comment]
@carloruiz I noticed this bit of guidance in the docs (https://www.cockroachlabs.com/docs/stable/node-shutdown?filters=decommission#decommissioning): 

> to avoid possible disruptions in query performance, we recommend manually draining before decommissioning. For more information, see [Perform node shutdown](https://www.cockroachlabs.com/docs/stable/node-shutdown?filters=decommission#perform-node-shutdown).

So I think the correct thing to do here is to `drain` first, then `decommission`.

I think there was an oversight that was introduced in 7cec096b801176d65db1d68b863969ed2aaabee7. This step of `drain` releases sql instances: https://github.com/cockroachdb/cockroach/blob/563aa774ba21479ff0c48458929ba851588b66e4/pkg/server/drain.go#L461-L470

However, the logs from this example show that returning the `err` prevents the graceful shutdown from continuing.
```
W231005 16:55:17.521778 25965 sql/catalog/lease/storage.go:238 ⋮ [T1,Vsystem,n2] 435  error releasing lease ID=54 ver=1 expiration={2023-10-05 16:59:37.781883
 +0000 UTC}: txn exec: context canceled
W231005 16:55:17.521798 25965 sql/catalog/lease/storage.go:238 ⋮ [T1,Vsystem,n2] 436  error releasing lease ID=32 ver=1 expiration={2023-10-05 16:59:32.812227
 +0000 UTC}: txn exec: context canceled
W231005 16:55:17.521816 25965 sql/catalog/lease/storage.go:238 ⋮ [T1,Vsystem,n2] 437  error releasing lease ID=31 ver=1 expiration={2023-10-05 16:59:41.063774
 +0000 UTC}: txn exec: context canceled
W231005 16:55:17.521830 355 sql/sqlliveness/slinstance/slinstance.go:334 ⋮ [T1,Vsystem,n2] 438  exiting heartbeat loop
I231005 16:55:17.521838 355 sql/sqlliveness/slinstance/slinstance.go:319 ⋮ [T1,Vsystem,n2] 439  draining heartbeat loop
I231005 16:55:17.521871 25965 1@server/drain.go:319 ⋮ [T1,Vsystem,n2] 440  drain remaining: 53
I231005 16:55:17.521881 25965 1@server/drain.go:321 ⋮ [T1,Vsystem,n2] 441  drain details: descriptor leases: 53
E231005 16:55:17.521890 25965 1@server/drain.go:177 ⋮ [T1,Vsystem,n2] 442  drain failed: txn exec: context canceled
```

I think this can be fixed by logging the error rather than returning it.

**Comment by carloruiz** [comment]
Yea we have on[ open issue ](https://cockroachlabs.atlassian.net/browse/CC-9541?jql=text%20~%20%22drain%20decommission%22%20AND%20project%20IN%20(10010))to drain before decommission. That said though, it is not expected for the node to block on termination. Looking forward to the bug fix. I imagine it will be backported right?

**Comment by rafiss** [comment]
@carloruiz actually I was wrong about my above theory. I've been testing this more. From what I've been able to reproduce, I believe this problem only occurs if you run `cockroach node decommission <node-id>` and the `node-id` being decommissioned is the same as the `node-id` that you send the decommission command to. Does that line up with your experience?

If so, then I don't think this can be fixed.

If you do try that, the following warning will appear on the CLI where you run `node decommission`:
```
W231010 17:12:21.916036 1 cli/node.go:604  [-] 2  skipping drain step for node n1; it is decommissioning and serving the request
```

The problem is that certain shutdown steps are skipped in this case. This is the same reason that the `--self` flag of `node decommission` was deprecated.

---

I was able to successfully decommission then SIGTERM in all my other tests where the decommission command was not sent to the node being decommissioned.

**Comment by carloruiz** [comment]
> This is the same reason that the --self flag of node decommission was deprecated.

Do you have a link that details why certain shutdown steps are skipped ?

**Comment by rafiss** [comment]
This is the relevant PR: https://github.com/cockroachdb/cockroach/pull/74319

The comment here explains why: https://reviewable.io/reviews/cockroachdb/cockroach/74319#-Ms8MpE1_lmNLbG3uMiH-r1-486:-MtEZn7YB3UXZ8Dropcl:b-wjz6ep

Basically, the final request to mark the node as "decommissioned" (and move out of the "decommissioning" state) cannot work if the node receiving that request has already been drained. That's because draining prevents all RPCs from working.

And on the flip-side, once a node has been fully decommissioned, none of the operations that are performed during the draining process can work, since the node can no longer update anything in storage. So graceful shutdown in this case is not an option, and instead the node must be killed non-gracefully, either by sending SIGKILL or by sending SIGINT twice.

**Comment by carloruiz** [comment]
Thanks for the link and info. This still sounds like a CRDB issue that needs fixing.

> That's because draining prevents all RPCs from working

Do you not see this as problematic? Seems undesirable to drain all SQL and grpc connections simultaneously since it will cause the issue you describe above. Could we instead:
1. Drain all SQL connections
2. Mark as decommissioning (maybe 1-2 can be flipped)
3. Shed leases
4. Transfer away replicas (at this point the node is a ghost node. IIUC, interrupting node-node coms should not impact SQL statements running in the cluster in any way).
5. Mark itself as decommissioned.


> A follow-up PR will fix this by enforcing that nodes serving requests are not target nodes. Decommission requests will be routed to other nodes in the cluster that are not targets

I understand the alternative I mention above may mean a significant refactor, so we could settle for what Cameron proposed in the PR. This would be a big improvement in operator experience. When you want to decommission a node, it is natural for the operator to go to _that_ node to issue to decommission command

**Comment by rafiss** [comment]
I think I can see the argument for why this could be considered a bug. However, in this case there also is a very clear and simple workaround - send the drain request before sending the decommission request. Given that, it's hard to justify working on the bug fix when we have many competing priorities. So far, our external customers have been satisfied by the guidance to do the drain first. Do you have any thoughts on why this workaround is not acceptable?

**Comment by carloruiz** [comment]
Ah, it was unclear that sending the drain command first would also mitigate the issue. If that's the case, that works for us and I can consider our work unblocked.

> A follow-up PR will fix this by enforcing that nodes serving requests are not target nodes. Decommission requests will be routed to other nodes in the cluster that are not targets

This is still worth doing IMO as it seems very low effort and will improve operator experience.

**Comment by rafiss** [comment]
I paired with Carlo on this a bit more. Some of this is summarized in https://github.com/cockroachlabs/managed-service/pull/14989, but a takeaway is that we will likely need to either:
- add a non-blocking drain command, OR
- make decommission drain the node (or perhaps even forcefully drain) even if it's running on the node being decommissioned.

**Comment by rafiss** [comment]
A workaround for this was found in https://github.com/cockroachlabs/managed-service/pull/14989, so i'm closing this as won't fix.

**Comment by carloruiz** [comment]
The workaround does not always work. It is unclear why sometime the termination completes successfully while at other times it hangs. Additionally, issuing `decommission` from another node is not equivalent to running `cockroach drain` followed by `cockroach decommission` on the same node.
```
-[ RECORD 6 ]
id                 | 6
address            | cockroachdb-7b9vg.cockroachdb.us-east-1.svc.cluster.local:26258
sql_address        | cockroachdb-7b9vg.cockroachdb.us-east-1.svc.cluster.local:26257
build              | v23.1.7
started_at         | 2023-11-06 22:29:24.933272 +0000 UTC
updated_at         | 2023-11-06 22:34:29.981064 +0000 UTC
locality           | region=aws-us-east-1,az=aws-us-east-1c,dns=cockroachdb-7b9vg.cockroachdb.us-east-1.svc.cluster.local
is_available       | false
is_live            | false
gossiped_replicas  | 0
is_decommissioning | true
membership         | decommissioned
is_draining        | false
```

**Comment by pritesh-lahoti** [comment]
@carloruiz , @rafiss is there anything else that we could try out here?

We have been trying to use the cloud operator for self-hosted k8s deployments, and the node resources are not being relinquished as part of teardown.

cc - @jmcarp

**Comment by rafiss** [comment]
The code here implements the logic of moving replicas away, then draining, and the fully decommissioning the node: https://github.com/cockroachdb/cockroach/blob/0ef13b670cb268ab7e6f55ca25536b4387d5dc2f/pkg/server/decommission.go#L311-L391

If there is any unavailability still being caused by a decommission operation, we'd need more precise reproduction instructions. The KV team would be best able to handle issues of unavailability during decommission.
